peer/peer.go

```diff
@@ -89,38 +93,99 @@ func New(ip string, port int) (*Peer, error) {
 }
 
 // Receive is the function that gets called when a remote peer
-// opens a new stream with the host that SetStreamHandler() is called on.
-// This should be passed as the second argument to SetStreamHandler().
-// We may want to implement another type of StreamHandler in the future.
+// opens a new stream with this peer.
+// This should be passed as the second argument to SetStreamHandler() after this
+// peer is initialized.
 func (p *Peer) Receive(s net.Stream) {
-	log.Debug("Setting basic stream handler.")
-	p.doCumulus(s)
-}
+	defer s.Close()
+
+	// Get remote peer's full multiaddress
+	remoteMA, err := makeMultiaddr(
+		s.Conn().RemoteMultiaddr(), s.Conn().RemotePeer())
+	if err != nil {
+		log.Fatal("Failed to obtain valid remote peer multiaddress")
+	}
+
+	// Add the remote peer to this peer's subnet
+	err = p.subnet.AddPeer(remoteMA, s)
```

Comment:
`AddPeer` handles the case where the peer is already in our subnet. I think your description of what `Receive` should do hits the nail on the head.